### PR TITLE
feat: support --no-optional on install process

### PR DESCRIPTION
**Pull request by fengmk2**
Don’t install optional dependencies.

like yarn does: https://yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-ignore-optional

### CODE REVIEW

File: lib/local_install.js [hunk 1]
```diff
@@ -147,9 +147,12 @@ function* _install(options) {
   options.spinner && options.spinner.succeed(`Installed ${tasks.length} packages`);
 
   if (!options.disableDedupe) {
-    // dedupe mode https://docs.npmjs.com/cli/dedupe
-    // link every packages' latest version to target directory
-    yield linkAllLatestVersion(rootPkgsMap, options);
+    // don't link latest versions on `npminstall pkgName` way
+    if (options.pkgs.length === 0) {
```
Comment (fengmk2) [inline lib/local_install.js]: 之前的实现有 bug

File: lib/dependencies.js [hunk 2]
```diff
@@ -32,8 +32,13 @@ module.exports = function dependencies(pkg) {
   }
   // follow npm, optionalDependencies will rewrite dependencies
   for (const name in optionalDependencies) {
-    all[name] = optionalDependencies[name];
-    prod[name] = optionalDependencies[name];
+    if (options.ignoreOptionalDependencies) {
+      delete all[name];
```
Comment (dead-horse) [inline lib/dependencies.js]: 如果在 dependencies 和 optionalDependencies 中都有，也会被删掉么？npm 是这样处理的？

Comment (fengmk2) [reply lib/dependencies.js]: npm 叫 --no-optional

**Comment by fengmk2** [comment]
yarn 叫 ignore-optional
npm 是 --no-optional ...

**Comment by dead-horse** [comment]
node 4 挂了

**Comment by fengmk2** [comment]
3.16.0

**Pull request merged**